qutip/solver/flimesolve.py

```diff
@@ -59,107 +59,103 @@ def _floquet_rate_matrix(floquet_basis,
-        rate_products_idx = np.argwhere(abs(rate_products) != 0)
-
-        args_key = list(omega.keys())[0]
-        valid_indices = [tuple(indices) for indices in rate_products_idx
-                         if delta(*tuple(indices)) == 0
-                         or abs((rate_products[tuple(indices)]
-                                 / (omega[args_key]
-                                    * delta(*tuple(indices))))**(-1)) <= time_sense
-                         and abs(omega[args_key] * delta(*tuple(indices))) < Nt / 2]
-
-        delta_dict = {}
-        for indices in valid_indices:
-            try:
-                delta_dict[delta(*indices)].append(indices)
-            except KeyError:
-                delta_dict[delta(*indices)] = [indices]
-
-        for key in delta_dict.keys():
-            mask_array = np.zeros((Hdim, Hdim, Hdim, Hdim, Nt, Nt))
-            for indices in delta_dict[key]:
-                mask_array[indices] = True
-            valid_c_op_products = rate_products * mask_array
-
-            # using c = ap,d = bp,k=lp
-            flime_FirstTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ma,nc,pb,qd->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-
-            flime_SecondTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ac,md,pb,qn->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-
-            flime_ThirdTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ac,nb,pm,qd->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-            try:
-                total_R_tensor[key] += np.reshape(flime_FirstTerm - (1 / 2)
-                                                  * (flime_SecondTerm +
-                                                     flime_ThirdTerm),
-                                                  (Hdim**2, Hdim**2))
-            except KeyError:
-                total_R_tensor[key] = np.reshape(flime_FirstTerm - (1 / 2)
-                                                 * (flime_SecondTerm +
-                                                    flime_ThirdTerm),
-                                                 (Hdim**2, Hdim**2))
+
+        for l, k in product(np.arange(Nt), repeat=2):
+            delta_shift = delta_m + (l - k)
+            mask = {}
+            if time_sense <= 0.:
+                mask[0] = delta_shift == 0
```

Comment:
Loop in python are slow, a for loop over an array is 100x slower than using a numpy function.
I kept the exact `== 0` but we are working with floating point, which can have surprising limitation, for example order of operation in addition can change the result:`(a + b) + c != a + (b +c)`. So this could miss some values that should have been included. Maybe (`delta_shift <= tol`) with `tol ~ 1e-14` would be better.